internal/ls/hover.go

```diff
@@ -52,143 +26,367 @@ func (l *LanguageService) ProvideHover(ctx context.Context, documentURI lsproto.
+		b.WriteString("import ")
+		b.WriteString(c.SymbolToStringEx(symbol, nil, ast.SymbolFlagsNone, symbolFormatFlags))
+	default:
+		b.WriteString(c.TypeToStringEx(c.GetTypeOfSymbol(symbol), nil, typeFormatFlags))
+	}
+	return b.String(), declaration
+}
+
+func inConstructorContext(node *ast.Node) bool {
+	if node.Kind == ast.KindConstructorKeyword {
+		return true
+	}
+	if ast.IsIdentifier(node) {
+		for ast.IsRightSideOfQualifiedNameOrPropertyAccess(node) {
+			node = node.Parent
+		}
+		if ast.IsNewExpression(node.Parent) {
+			return true
+		}
+	}
+	return false
+}
+
+func writeTypeParams(b *strings.Builder, c *checker.Checker, params []*checker.Type) {
+	if len(params) > 0 {
+		b.WriteString("<")
+		for i, tp := range params {
+			if i != 0 {
+				b.WriteString(", ")
 			}
-		case flags&(ast.SymbolFlagsClass|ast.SymbolFlagsInterface) != 0:
-			b.WriteString(core.IfElse(symbol.Flags&ast.SymbolFlagsClass != 0, "class ", "interface "))
-			b.WriteString(c.SymbolToStringEx(symbol, file.AsNode(), ast.SymbolFlagsNone, symbolFormatFlags))
-			params := c.GetDeclaredTypeOfSymbol(symbol).AsInterfaceType().LocalTypeParameters()
-			writeTypeParams(params, c, file, &b)
-		case flags&ast.SymbolFlagsEnum != 0:
-			b.WriteString("enum ")
-			b.WriteString(c.SymbolToStringEx(symbol, file.AsNode(), ast.SymbolFlagsNone, symbolFormatFlags))
-		case flags&ast.SymbolFlagsModule != 0:
-			b.WriteString(core.IfElse(symbol.ValueDeclaration != nil && ast.IsSourceFile(symbol.ValueDeclaration), "module ", "namespace "))
-			b.WriteString(c.SymbolToStringEx(symbol, file.AsNode(), ast.SymbolFlagsNone, symbolFormatFlags))
-		case flags&ast.SymbolFlagsTypeParameter != 0:
-			b.WriteString("(type parameter) ")
-			tp := c.GetDeclaredTypeOfSymbol(symbol)
-			b.WriteString(c.SymbolToStringEx(symbol, file.AsNode(), ast.SymbolFlagsNone, symbolFormatFlags))
+			symbol := tp.Symbol()
+			b.WriteString(c.SymbolToStringEx(symbol, nil, ast.SymbolFlagsNone, symbolFormatFlags))
 			cons := c.GetConstraintOfTypeParameter(tp)
 			if cons != nil {
 				b.WriteString(" extends ")
-				b.WriteString(c.TypeToStringEx(cons, file.AsNode(), typeFormatFlags))
+				b.WriteString(c.TypeToStringEx(cons, nil, typeFormatFlags))
+			}
+		}
+		b.WriteString(">")
+	}
+}
+
+func writeSignatures(b *strings.Builder, c *checker.Checker, signatures []*checker.Signature, prefix string, symbol *ast.Symbol) {
+	for i, sig := range signatures {
+		if i != 0 {
+			b.WriteString("\n")
+		}
+		if i == 3 && len(signatures) >= 5 {
+			b.WriteString(fmt.Sprintf("// +%v more overloads", len(signatures)-3))
+			break
+		}
+		b.WriteString(prefix)
+		b.WriteString(c.SymbolToStringEx(symbol, nil, ast.SymbolFlagsNone, symbolFormatFlags))
+		b.WriteString(c.SignatureToStringEx(sig, nil, typeFormatFlags|checker.TypeFormatFlagsWriteCallStyleSignature))
+	}
+}
+
+func formatQuickInfoAndJSDoc(quickInfo string, declaration *ast.Node) string {
+	var b strings.Builder
+	b.Grow(32)
+	writeCode(&b, "ts", quickInfo)
```

Comment:
Should this be `tsx` to match the old tooltips?

Reply:
Yeah, I'll fix.